packages/xo-server/src/index.js

```diff
@@ -140,6 +147,50 @@ async function setUpPassport(express, xo) {
     res.redirect('/')
   })
 
+  express.get('/signin/otp', (req, res, next) => {
+    if (req.session.user === undefined) {
+      return res.redirect('/signin')
+    }
+
+    res.send(
+      signInPage({
+        error: req.flash('error')[0],
+        otp: true,
+        strategies,
+      })
+    )
+  })
+
+  express.post('/signin/otp', (req, res, next) => {
+    const { user } = req.session
+
+    if (user === undefined) {
+      return res.redirect('/signin')
+    }
+
+    if (authenticator.check(req.body.otp, user.preferences.otp)) {
+      setToken(req, res, next)
+    } else {
+      req.flash('error', 'Invalid code')
+      return res.redirect(303, '/signin/otp')
+    }
+  })
+
+  const setToken = async (req, res, next) => {
+    const { user, isPersistent } = req.session
+    const token = (await xo.createAuthenticationToken({ userId: user.id })).id
+
+    // Persistent cookie ? => 1 year
+    // Non-persistent : external provider as Github, Twitter...
+    isPersistent === true
+      ? res.cookie('token', token, { maxAge: 1000 * 60 * 60 * 24 * 365 })
+      : res.cookie('token', token)
```

Comment:
I prefer reserving the use of the ternary operator when the expression value is used:

```js
res.cookie('token', token, isPersistent ? { maxAge: 1000 * 60 * 60 * 24 * 365 } : undefined)
```